### ISSUE TITLE
Setting max.term.length property does not work with IterDictIndexer

### ISSUE THREAD

**Issue by ad4529** [terrier-org/pyterrier#499]
**Describe the bug**
I have want to index a file of large words using IterDictIndexer by loading a TSV file into a pandas DataFrame and building an index. However, using the IterDictIndexer does not work as likely this indexer is parallelized and the config is not passed onto the worker threads. I took a look at the issue described here https://github.com/terrier-org/pyterrier/issues/357 but the solution did not work for this indexer.

I tried setting the properties

**To Reproduce**
This is an excerpt of the TSV file containing SMILES strings:
```
docno	text	IUPAC
120	CCCN/C=N/C(=N/c1c(C#N)c[nH]c1/N=C/Nc1ccc(OC)cc1OC)Nc1ccc(S(=O)(=O)N2CCC(N3CCOCC3)CC2)cc1	N'-[4-cyano-2-[(E)-(2,4-dimethoxyanilino)methylideneamino]-1H-pyrrol-3-yl]-N-[4-(4-morpholin-4-ylpiperidin-1-yl)sulfonylphenyl]-N-propylmethanimidamide
121	C=CC(=CC=COC)S(=O)(=O)N1C=C(OC)C2(NC3=NC4=C(C(C#N)=CN4)C(NCC(C)C)=N3)=CC=21	2-[[4-methoxy-2-(6-methoxyhexa-1,3,5-trien-3-ylsulfonyl)-2-azabicyclo[4.1.0]hepta-3,5-dien-6-yl]amino]-4-(2-methylpropylamino)-7H-pyrrolo[2,3-d]pyrimidine-5-carbonitrile
122	CNc1ccc(S(=O)(=O)N2CC=C(N/C=N/C3=C(C(C)C)C(NC#N)=N3)CC2)cc1OC	N'-[4-(cyanoamino)-3-propan-2-ylazet-2-yl]-N-[1-[3-methoxy-4-(methylamino)phenyl]sulfonyl-3,6-dihydro-2H-pyridin-4-yl]methanimidamide
```

Minimal reproducible code:

```python
import pandas as pd
import pyterrier as pt

def rdkit_search(smi_path):
    
    # Init Pyterrier
    if not pt.started():
        pt.init()
        pt.set_property("max.term.length", "500")
    
    # Load the pandas dataframe
    df_smi = pd.read_csv(smi_path, sep='\t')
   
    # Convert any ints to floats
    int_columns = df_smi.select_dtypes(include='int64').columns
    df_smi[int_columns] = df_smi[int_columns].astype(str)
    df_smi.columns = ['docno', 'text']

   # Create and save the indexes
    indexer_par = pt.IterDictIndexer('./index_store/smi_index', overwrite=True, verbose=True, 
                                 stemmer=None, stopwords=None, tokeniser='identity',
                                 meta={'docno':20, 'text':500}, properties={
                                     "max.term.length": "500",
                                     "indexer.meta.forward.keys": "500",
                                     "indexer.meta.forward.keylens": "500",
                                     "termpipelines": ""})

    indexref_par = indexer_par.index(df_smi.to_dict(orient='records'))
    index_par = pt.IndexFactory.of(indexref_par)

    for kv in index_par.getLexicon():
        print(f'{kv.getKey():40.40} {kv.getValue().toString():50.50}')


if __name__ == '__main__':
    df_smi_path = '<INSERT PATH HERE TO THE TSV>'
    rdkit_search(df_smi_path)
```

**Error**

```shell
22:27:56.767 [ForkJoinPool-1-worker-1] ERROR org.terrier.structures.seralization.FixedSizeTextFactory -- TermCCCN/C=N/C(=N/c1c(C#N)c[nH]c1/N=C/Nc1ccc(OC)cc1OC)Nc1ccc(S(=O)(=O)N2CCC(N3CCOCC3)CC2)cc1 written in 89 bytes. Max expected size was 61. Increase relevant property: max.term.length for Lexicon, or indexer.meta.forward.keylens for metadata
java.lang.ArrayIndexOutOfBoundsException: arraycopy: length -28 is negative
	at java.base/java.lang.System.arraycopy(Native Method)
	at java.base/java.io.BufferedOutputStream.implWrite(BufferedOutputStream.java:224)
```

**Documentation and Issues**
 - [*] I have checked the [PyTerrier documentation](https://pyterrier.readthedocs.io/en/latest/) for relevant content
 - [ *] I have checked for previous relevant [PyTerrier issues](https://github.com/terrier-org/pyterrier/issues)

**Additional context**
Any workaround to this would be appreciated. I ideally do not want to convert my data into the TREC format.

Thank you!
Abhisek

**Comment by cmacdonald** [comment]
Thanks for the report @ad4529, I'll try to get a short test case for this today so I can debug the problem

**Comment by cmacdonald** [comment]
There are a couple of things going on:
- properties kwarg object seems to be lost.
- I need to check when ApplicationSetup.MAX_TERM_LENGTH is used or updated.

This looks like a workaround:
```python
pt.java.set_log_level('DEBUG')
indexer = pt.IterDictIndexer("./index", tokeniser='identity', overwrite=True, properties={'max.term.length' : '80'})
indexer.setProperty('max.term.length', '79')
ref = indexer.index(docs)
```
with output
```
12:41:34.021 [ForkJoinPool-11-worker-3] INFO org.terrier.structures.indexing.Indexer -- Indexer using 1 fields
12:41:34.039 [ForkJoinPool-11-worker-3] INFO org.terrier.structures.indexing.Indexer -- creating the data structures data_stream0_1
12:41:34.047 [ForkJoinPool-11-worker-3] DEBUG org.terrier.structures.Index -- Files  /content/index/data_stream0_1.properties doesn't exist
12:41:34.059 [ForkJoinPool-11-worker-3] DEBUG org.terrier.structures.Index -- Creating new index : /content/index/data_stream0_1.properties
12:41:34.065 [ForkJoinPool-11-worker-3] DEBUG org.terrier.utility.RuntimeMemoryChecker -- memory.reserved=200 MiB memory.heap.usage=0.7
12:41:34.077 [ForkJoinPool-11-worker-3] DEBUG org.terrier.structures.Index -- Attempting to load structure lexicon-valuefactory
12:41:34.083 [ForkJoinPool-11-worker-3] INFO org.terrier.structures.indexing.LexiconBuilder -- LexiconBuilder active - flushing every 100000 documents, or when memory threshold hit
12:41:34.098 [ForkJoinPool-11-worker-3] DEBUG org.terrier.utility.RuntimeMemoryChecker -- memory.reserved=200 MiB memory.heap.usage=0.7
12:41:34.104 [ForkJoinPool-11-worker-3] DEBUG org.terrier.structures.indexing.BaseMetaIndexBuilder -- InitialisingZstdMetaIndexBuilder
12:41:34.115 [ForkJoinPool-11-worker-3] DEBUG org.terrier.structures.seralization.FixedSizeTextFactory -- FixedSizeTextFactory: maxKeyWrittenSize=61 for term of size 20
12:41:34.133 [ForkJoinPool-11-worker-3] DEBUG org.terrier.structures.indexing.BaseMetaIndexBuilder -- Finished writing metaindex: keys [docno] keylens [20] sorted [true] data file size 18
12:41:34.142 [ForkJoinPool-11-worker-3] INFO org.terrier.structures.indexing.BaseMetaIndexBuilder -- ZstdMetaIndexBuilder meta achieved compression ratio 3.3888888 (> 1 is better)
12:41:34.149 [ForkJoinPool-11-worker-3] DEBUG org.terrier.structures.indexing.LexiconBuilder -- flushing lexicon to disk after the direct index completed
12:41:34.160 [ForkJoinPool-11-worker-3] DEBUG org.terrier.structures.Index -- Attempting to load structure lexicon-keyfactory
12:41:34.181 [ForkJoinPool-11-worker-3] DEBUG org.terrier.structures.seralization.FixedSizeTextFactory -- FixedSizeTextFactory: maxKeyWrittenSize=238 for term of size 79
12:41:34.189 [ForkJoinPool-11-worker-3] INFO org.terrier.structures.indexing.LexiconBuilder -- 1 lexicons to merge
```

Given the format of your documents, I'd recommend pre-tokenised format, for more direct control over the tokens, but the same bug would probably exist.

**Comment by cmacdonald** [comment]
We've addressed this in #500, which will be in the next PyTerrier release. Thanks @ad4529 for the report.

**Comment by cmacdonald** [comment]
excellent. Hopefully your other tools get upgraded soon, we have more upgrades coming for PyTerrier and Python 3.7 is now unsupported.
